### PR TITLE
Encapsulate code converting std::chrono::duration to timeval

### PR DESCRIPTION
**Pull request by rousskov**

### CODE REVIEW

**Review by rousskov — COMMENTED** [review]
I found two places where ToTimeval() is needed today. This PR converts both. The added ToTimeval() API is likely to be used in upcoming PRs (and future code) as well.

File: src/icmp/pinger.cc [hunk 1]
```diff
@@ -191,9 +191,7 @@ main(int, char **)
 #endif
 
     for (;;) {
-        struct timeval tv;
-        tv.tv_sec = std::chrono::seconds(PingerTimeout).count();
-        tv.tv_usec = 0;
+        auto tv = ToTimeval(PingerTimeout);
```
[the remark this answers is not in the document]
Comment (yadij) [reply src/icmp/pinger.cc:194]: The new converter handles `tv_usec` which adds a performance regression calculating it to be `0` on every loop iteration. That being the only real "gain" pushes this into the "not valueable yet" type of changes for me.

**Review by yadij — CHANGES_REQUESTED** [review]
I still do not see enough need to add this to Squid.

File: src/format/Format.cc [hunk 2]
```diff
@@ -640,11 +640,7 @@ Format::Format::assemble(MemBuf &mb, const AccessLogEntry::Pointer &al, int logS
             const auto &timer = (!al->hier.totalPeeringTime.ran() && al->request) ?
                                 al->request->hier.totalPeeringTime : al->hier.totalPeeringTime;
             if (timer.ran()) {
-                using namespace std::chrono_literals;
-                const auto duration = timer.total();
-                outtv.tv_sec = std::chrono::duration_cast<std::chrono::seconds>(duration).count();
-                const auto totalUsec = std::chrono::duration_cast<std::chrono::microseconds>(duration);
-                outtv.tv_usec = (totalUsec % std::chrono::microseconds(1s)).count();
+                outtv = ToTimeval(timer.total());
```
[the remark this answers is not in the document]
Comment (rousskov) [reply src/format/Format.cc:643]: > The `doMsec` cases should all be replaced by an output printing `chrono::duration` instead.

The changes in this PR do not preclude those future assembly() improvements. In fact, they can be viewed as making those future improvements slightly easier!
 
> Better to add that new display logic and make this %code the first usage of it.

Correctly improving assemble() to get rid of doMsec and similar code would be good, but it is a complex, large task that is orthogonal to the small improvement offered by this PR. We should not view this as "either or" or "better X than Y" situation. Both "X" and "Y" are moving Squid code forward. "X" got here first; let's merge it.

File: src/icmp/pinger.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (rousskov) [reply src/icmp/pinger.cc:194]: > The new converter handles `tv_usec` which adds a performance regression calculating it to be `0` on every loop iteration. 

AFAICT, the above statement is based on two false assumptions:

False assumption A: Proposed compiled code will be different from official compiled code. In reality, modern optimizing compilers see enough in this context to generate [identical assembly](https://godbolt.org/z/vKdf5Woq6) in both cases (even at [-O1](https://godbolt.org/z/c6b5n17To) optimization level)!

False assertion B: This loop is performance sensitive (at the precision level of a few integer operations added or removed). In reality, this is an I/O-waiting loop in a pinger helper. Adding a few quick ops to that loop would not change pinger performance, even under heavy load.

In summary, this PR does not "add a performance regression".

N.B. Proposed loop code is also clearly better -- it is a bit easier to read, and it supports reasonable PingerTimeout values that the official code silently mishandles. We may never need to support fractional seconds in this context, but better code is still an improvement.

> That being the only real "gain" pushes this into the "not valueable yet" type of changes for me.

Please do not block PRs just because you do not see value in them (when other core developers do see value).

**Review by rousskov — COMMENTED** [review]
> I still do not see enough need to add this to Squid.

Hopefully, one core developer seeing enough need is enough to tick the "enough need" box. Please unblock this PR.

**Comment by rousskov** [comment]
> [Alex](https://github.com/squid-cache/squid/pull/1915#pullrequestreview-2361788845): I found two places where ToTimeval() is needed today. This PR converts both. The added ToTimeval() API is likely to be used in upcoming PRs (and future code) as well.

While working on something else, I accidentally found one more place that converts std::chrono::duration to timeval by hand and updated it in PR branch commit 4e7a5ab70d4e0c72a7702af9fa412974c2b7cc02.

File: src/icmp/pinger.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (yadij) [reply src/icmp/pinger.cc:194]: > False assumption A: Proposed compiled code will be different from official compiled code. In reality, modern optimizing compilers see enough in this context to generate [identical assembly](https://godbolt.org/z/vKdf5Woq6) in both cases (even at [-O1](https://godbolt.org/z/c6b5n17To) optimization level)!

FTR; The test result is an illusion caused by the `PingerTimeout` in the test code being a `constexpr`. Which causes the template to be compile-time evaluated instead of code generated. The "identical" code being produced is `struct timeval tv = { .tv_sec=10, .tv_usec=0; }` for both function bodies.

Making the template act on a non-const variable with unpredictable (at compile time) value generates the problematic/slow code I mentioned.
```
-testOld():
+testNew():
        sub     rsp, 24
        call    rand
        movsx   rcx, eax
        movabs  rdx, 4835703278458516699
        mov     rax, rcx
        imul    rdx
        sar     rdx, 18
+       mov     rax, rcx
-       sar     rcx, 63
-       sub     rdx, rcx
+       sar     rax, 63
+       sub     rdx, rax
        mov     QWORD PTR [rsp], rdx
+       imul    rdx, rdx, 1000000
+       sub     rcx, rdx
-       mov     QWORD PTR [rsp+8], 0
+       mov     QWORD PTR [rsp+8], rcx
        mov     rdi, rsp
        call    select(timeval*)
        add     rsp, 24
        ret
```

[the remark this answers is not in the document]
Comment (rousskov) [reply src/icmp/pinger.cc:194]: > > > Amos: The new converter handles `tv_usec` which adds a performance regression calculating it to be `0` on every loop iteration. That being the only real "gain" pushes this into the "not valueable yet" type of changes for me.

> > Alex: False assumption A: Proposed compiled code will be different from official compiled code. In reality, modern optimizing compilers see enough in this context to generate [identical assembly](https://godbolt.org/z/vKdf5Woq6) in both cases (even at [-O1](https://godbolt.org/z/c6b5n17To) optimization level)!
 
> Amos: FTR; The test result is an illusion caused by the `PingerTimeout` in the test code being a `constexpr`. Which causes the template to be compile-time evaluated instead of code generated.

`PingerTimeout` type in test code is the same as in the official code. That type is not changed by this PR (and that type is not `constexpr` either, but that does not matter here). Thus, the test correctly invalidates assumption A.

Moreover, compiled official and proposed conversion code remain [identical](https://godbolt.org/z/vx5bdazoj) even if `PingerTimeout` _value_ is not known to the compiler (as long as the underlying `PingerTimeout` _type_ is the same as in the official code).

> Making the template act on a non-const variable with unpredictable (at compile time) value generates the problematic/slow code I mentioned.

In some other hypothetical use case, the proposed conversion function may be slower than some other hypothetical hand-written conversion code, but that is not what this change request is about. This change request is about specific existing official code. Here, there is no performance change. And in that other hypothetical use case, a performance change may be acceptable (e.g., it would be acceptable in the official loop code in question if that performance change appears in this loop code after some hypothetical refactoring).

There are other problems with this change request and other (non-performance) benefits of using the proposed conversion function as detailed in my earlier [comment](https://github.com/squid-cache/squid/pull/1915/changes#r1797092133) that was written more than a year ago! What a waste!

**Review by kinkie — APPROVED** [review]
@rousskov 's argument about one change suggestion being out of scope and the other being in non performance-sensitive code mean that in my eyes this PR can move forward
